File: apps/android/app/src/main/java/com/accessmate/data/DraftBuilder.kt
package com.accessmate.data

import java.time.Instant

/** Mirror of `assembleDraft` in packages/shared and iOS DraftBuilder. */
object DraftBuilder {
    fun assemble(
        template: ScenarioTemplate,
        facts: IncidentFacts,
        op: OperatorEntry?,
        profile: Profile?,
    ): String {
        val date = (facts.whenISO ?: Instant.now().toString()).take(10)
        val operatorName = op?.name ?: facts.operatorName ?: "the operator"

        fun fill(s: String): String =
            s.replace("{{operator}}", operatorName)
                .replace("{{date}}", date)
                .replace("{{location}}", "the location described above")

        val about = buildList {
            when (profile?.mobility?.wheelchairType) {
                "powered" -> add("- Powered wheelchair user.")
                "manual" -> add("- Manual wheelchair user.")
                "mobility-scooter" -> add("- Mobility scooter user.")
            }
            if (profile?.sensory?.isBlind == true) add("- Blind.")
            if (profile?.sensory?.isLowVision == true) add("- Low vision.")
            if (profile?.sensory?.isDeaf == true) add("- Deaf.")
            if (profile?.sensory?.isHardOfHearing == true) add("- Hard of hearing.")
            if (profile?.communication?.prefersBSL == true) add("- Prefers British Sign Language.")
            if (profile?.communication?.prefersWriting == true) add("- Prefers written communication.")
            if (profile?.communication?.needsExtraTime == true) add("- Needs extra time.")
        }

        val aboutMe = if (about.isEmpty()) {
            "No specific accessibility profile shared for this complaint."
        } else {
            about.joinToString("\n")
        }

        val factLines = buildList {
            add("- Date: $date.")
            add("- Operator: $operatorName.")
            when (facts.accompanied) {
                true -> add("- Travelling with a companion.")
                false -> add("- Travelling alone.")
                null -> Unit
            }
            facts.staffInteractions?.takeIf { it.isNotBlank() }?.let { add("- Staff interactions: $it.") }
            facts.witnesses?.takeIf { it.isNotBlank() }?.let { add("- Witnesses: $it.") }
            facts.waitedMinutes?.let { add("- Waited approximately $it minutes.") }
            facts.narrative?.takeIf { it.isNotBlank() }?.let { add("- Summary: $it.") }
        }

        return listOf(
            "# ${template.title}",
            "",
            fill(template.header),
            "",
            "## What happened",
            "",
            factLines.joinToString("\n"),
            "",
            "## About me",
            "",
            aboutMe,
            "",
            "## Legal context",
            "",
            fill(template.legalParagraph),
            "",
            "## What I want",
            "",
            fill(template.ask),
            "",
        ).joinToString("\n")
    }
}

File: apps/ios/AccessMate/Data/DraftBuilder.swift
import Foundation

/// Mirror of `assembleDraft` in `packages/shared/src/index.ts`. Every
/// platform must produce the same draft body for the same inputs.
enum DraftBuilder {
    static func assemble(
        template: ScenarioTemplate,
        facts: IncidentFacts,
        op: OperatorEntry?,
        profile: Profile?
    ) -> String {
        let date = String((facts.whenISO ?? ISO8601DateFormatter().string(from: Date())).prefix(10))
        let operatorName = op?.name ?? facts.operatorName ?? "the operator"

        func fill(_ s: String) -> String {
            s
                .replacingOccurrences(of: "{{operator}}", with: operatorName)
                .replacingOccurrences(of: "{{date}}", with: date)
                .replacingOccurrences(of: "{{location}}", with: "the location described above")
        }

        var aboutLines: [String] = []
        if let m = profile?.mobility {
            switch m.wheelchairType {
            case "powered":           aboutLines.append("- Powered wheelchair user.")
            case "manual":            aboutLines.append("- Manual wheelchair user.")
            case "mobility-scooter":  aboutLines.append("- Mobility scooter user.")
            default: break
            }
        }
        if profile?.sensory?.isBlind == true            { aboutLines.append("- Blind.") }
        if profile?.sensory?.isLowVision == true        { aboutLines.append("- Low vision.") }
        if profile?.sensory?.isDeaf == true             { aboutLines.append("- Deaf.") }
        if profile?.sensory?.isHardOfHearing == true    { aboutLines.append("- Hard of hearing.") }
        if profile?.communication?.prefersBSL == true   { aboutLines.append("- Prefers British Sign Language.") }
        if profile?.communication?.prefersWriting == true { aboutLines.append("- Prefers written communication.") }
        if profile?.communication?.needsExtraTime == true { aboutLines.append("- Needs extra time.") }

        let aboutMe = aboutLines.isEmpty
            ? "No specific accessibility profile shared for this complaint."
            : aboutLines.joined(separator: "\n")

        var factLines: [String] = [
            "- Date: \(date).",
            "- Operator: \(operatorName).",
        ]
        if facts.accompanied == true  { factLines.append("- Travelling with a companion.") }
        if facts.accompanied == false { factLines.append("- Travelling alone.") }
        if let s = facts.staffInteractions, !s.isEmpty { factLines.append("- Staff interactions: \(s).") }
        if let w = facts.witnesses, !w.isEmpty          { factLines.append("- Witnesses: \(w).") }
        if let m = facts.waitedMinutes                  { factLines.append("- Waited approximately \(m) minutes.") }
        if let n = facts.narrative, !n.isEmpty          { factLines.append("- Summary: \(n).") }

        return [
            "# \(template.title)",
            "",
            fill(template.header),
            "",
            "## What happened",
            "",
            factLines.joined(separator: "\n"),
            "",
            "## About me",
            "",
            aboutMe,
            "",
            "## Legal context",
            "",
            fill(template.legalParagraph),
            "",
            "## What I want",
            "",
            fill(template.ask),
            "",
        ].joined(separator: "\n")
    }
}
